src/components/building/building.js

```diff
@@ -40,78 +41,57 @@ export const Building = () => {
-    gameBuildingType.buildMaterials.forEach(buildMaterial => {
-      for (const key in currentUserResources) {
-        if (key === buildMaterial.resourceType) {
-          
-          const currentResource = currentUserResources[key];
-          const neededResource = buildingPrice(nextLevel, gameBuildingType.id, key);
-          const enoughRescource = currentResource >= neededResource;
-          
-          const textColor = enoughRescource ? 'black' : 'red';
-          if (!enoughRescource) {
-            notEnoughRescourceButtonDisable = true;
-          }
-          buildPriceTextColor.push(textColor);
-        }
-      }
-    });
-
-    return (
-      <article key={gameBuildingType.id}>
-        <Buildingcard
-          buildType={0}
-          buildId={gameBuildingType.id}
-          buildName={gameBuildingType.name}
-          buildDescription={gameBuildingType.description}
-          buildTime={buildTime}
-          buildPrice={gameBuildingType.buildMaterials}
-          buildPriceTextColor={buildPriceTextColor}
-          buildYield={gameBuildingType.productionMaterials}
-          buildInProgressTime={buildInProgressTime}
-          currentBuildLevel={currentUserBuilding.level}
-          buttonText={buttonTexts}
-          buttonDisabled={notEnoughRescourceButtonDisable || buildInProgressButtonDisable }
-          buttonFunction={onHandleClickUpgrade}
-          currentUserBuildings={currentUserBuildings}
-        />
-      </article>
-    );
+  if (currentUserBuildingInProgress.length > 0) {
+    const inProgressId = currentUserBuildingInProgress[0].buildingId;
+    buttonTexts = inProgressId === gameBuildingType.id ? 'in Progress' : buttonTexts;
+  }
+
+  buttonTexts = currentUserBuilding.level > 0 && buttonTexts !== 'in Progress' ? 'Upgrade' : buttonTexts;
+
+  const nextLevel = currentUserBuilding.level + 1;
+
+  let buildPrice = [];
+  gameBuildingType.buildMaterials.forEach(material => {
+    if (material.resourceType === 'money' || material.resourceType === 'iron') {
+      buildPrice.push(buildingPrice(nextLevel, gameBuildingType.id, material.resourceType));
+    }
   });
+  const buildTime = buildingTime(
+    buildPrice[0] !== undefined ? buildPrice[0] : 0,
+    buildPrice[1] !== undefined ? buildPrice[1] : 0
+  );
+
+  const buildInProgressButtonDisable = currentUserBuildingInProgress.length === 0 ? false : true;
+
+  const checkEnoughRescource = checkRescource(gameBuildingType, currentUserResources, nextLevel);
+
+  return (
+    <article>
+      <Buildingcard
+        buildType={0}
+        buildId={gameBuildingType.id}
+        buildName={gameBuildingType.name}
+        buildDescription={gameBuildingType.description}
+        buildTime={buildTime}
+        buildPrice={gameBuildingType.buildMaterials}
+        buildPriceTextColor={checkEnoughRescource.buildPriceTextColor}
+        buildYield={gameBuildingType.productionMaterials}
+        currentBuildLevel={currentUserBuilding.level}
+        buttonText={buttonTexts}
```

Comment:
`buttonTexts` is a string and not an array, right? Then it would be better to name it in singular: `buttonText`.
Maybe it would be cleaner to pass only a state like `inProgress` to the `BuildingCard` component and let the component handle the Button state and text.

Reply:
that with the button text, there I just on the fast no suitable simple solution, there I would lose with thinking too much time. I would like to leave that so for now as long as it works so.

as soon as there is some free space i will adjust it, because more than half of the code can be outsourced anyway.